parlai/core/worlds.py

```diff
@@ -974,6 +1004,216 @@ def shutdown(self):
         self.world.shutdown()
 
 
+class DynamicBatchWorld(World):
+    def __init__(self, opt: Opt, world: Union[DialogPartnerWorld, MultiWorld]):
+        super().__init__(opt)
+        self.opt = opt
+
+        # agents is a placeholder just for super.reset
+        self.agents = []
+
+        # check some assumptions
+        if isinstance(
+            world, (ExecutableWorld, BatchWorld, HogwildWorld, MultiAgentDialogWorld)
+        ):
+            raise TypeError(
+                'World must be a DialogPartnerWorld or a '
+                'MultiWorld of DialogPartnerWorld'
+            )
+
+        if len(world.get_agents()) != 2:
+            raise AssertionError(
+                "Dynamic batch only works in a fixed dialog world with two agents."
+            )
+
+        if not hasattr(world.get_model_agent(), 'batch_act'):
+            raise TypeError("Model agent doesn't have batch_act.")
+
+        self.truncate = opt.get('text_truncate', None) or opt.get('truncate', None)
+        self.l_truncate = opt.get('label_truncate', None) or opt.get('truncate', None)
+        if self.truncate is None or self.truncate < 0:
+            raise ValueError(
+                'You must use --text-truncate or --truncate in order to use '
+                'dynamic batching.'
+            )
+
+        # size of the buffer we will use to find worlds
+        self._BUFFER_SIZE = 1021  # chosen as a prime number
+
+        if opt['dynamic_batching'] == 'full':
+            # full dynamic batching, we can grow our batchsize
+            self.max_batch_size = self._BUFFER_SIZE
+        else:
+            # simple batchsort
+            self.max_batch_size = opt['batchsize']
+
+        # TODO: check to ensure the agent has self_observe
+        shared = world.share()
+        self.world = world
+        # TODO: maybe generalize this
+        self.max_words = (self.l_truncate + self.truncate) * opt['batchsize']
+
+        # buffer worlds
+        self.worlds = [
+            shared['world_class'](opt, None, shared) for _ in range(self._BUFFER_SIZE)
```

Comment:
you mentioned this in my pr - do we want to maintain this `(opt, None, shared)` initialization?

Reply:
agreed, will change to match yours